File: code/cpp/ch27/learn/example_01_dijkstra_demo.cpp
/*
 * Example 01: Dijkstra's Algorithm — Step-by-Step Demo
 * Chapter 27: Shortest Paths — Finding the Best Route
 */

#include <algorithm>
#include <climits>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

vector<int> dijkstra(int n, vector<vector<int>>& edges, int src) {
    const int INF = 1e9;
    vector<vector<pair<int,int>>> adj(n);
    for (auto& e : edges)
        adj[e[0]].push_back({e[1], e[2]});

    vector<int> dist(n, INF);
    dist[src] = 0;
    priority_queue<pair<int,int>, vector<pair<int,int>>, greater<>> pq;
    pq.push({0, src});

    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (d > dist[u]) continue;
        cout << "  Process node " << u << " (distance = " << d << ")\n";
        for (auto [v, w] : adj[u]) {
            if (dist[u] + w < dist[v]) {
                cout << "    Relax " << u << " -> " << v
                     << ": " << dist[v] << " -> " << dist[u] + w << "\n";
                dist[v] = dist[u] + w;
                pq.push({dist[v], v});
            }
        }
    }
    return dist;
}

int main() {
    vector<vector<int>> edges = {{0,1,4},{0,2,1},{2,1,2},{1,3,5},{2,3,8},{3,4,1}};
    cout << "Dijkstra's Algorithm Demo\n";
    cout << "=========================\n";
    auto dist = dijkstra(5, edges, 0);
    cout << "Distances:";
    for (int d : dist) cout << " " << d;
    cout << "\n";
    return 0;
}

File: code/java/ch27/learn/Example01DijkstraDemo.java
package ch27.learn;

import java.util.*;

/**
 * Example 01: Dijkstra's Algorithm — Step-by-Step Demo
 * Chapter 27: Shortest Paths — Finding the Best Route
 */
public class Example01DijkstraDemo {

    static int[] dijkstra(int n, int[][] edges, int src) {
        int INF = (int) 1e9;
        List<int[]>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) adj[i] = new ArrayList<>();
        for (int[] e : edges) adj[e[0]].add(new int[]{e[1], e[2]});

        int[] dist = new int[n];
        Arrays.fill(dist, INF);
        dist[src] = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.offer(new int[]{0, src});

        while (!pq.isEmpty()) {
            int[] top = pq.poll();
            int d = top[0], u = top[1];
            if (d > dist[u]) continue;
            System.out.println("  Process node " + u + " (distance = " + d + ")");
            for (int[] edge : adj[u]) {
                int v = edge[0], w = edge[1];
                if (dist[u] + w < dist[v]) {
                    System.out.println("    Relax " + u + " -> " + v + ": " + dist[v] + " -> " + (dist[u] + w));
                    dist[v] = dist[u] + w;
                    pq.offer(new int[]{dist[v], v});
                }
            }
        }
        return dist;
    }

    public static void main(String[] args) {
        int[][] edges = {{0,1,4},{0,2,1},{2,1,2},{1,3,5},{2,3,8},{3,4,1}};
        System.out.println("Dijkstra's Algorithm Demo");
        System.out.println("=========================");
        int[] dist = dijkstra(5, edges, 0);
        System.out.println("Distances: " + Arrays.toString(dist));
    }
}
